crates/bevy_pbr/src/light_probe/environment_map.rs

```diff
@@ -176,16 +181,63 @@ pub struct EnvironmentMapViewLightProbeInfo {
     pub(crate) intensity: f32,
 }
 
-impl ExtractInstance for EnvironmentMapIds {
-    type QueryData = Read<EnvironmentMapLight>;
+#[derive(Resource, Default, Deref, DerefMut)]
+pub struct EnvironmentMapInstances(EntityHashMap<EnvironmentMapIds>);
 
-    type QueryFilter = ();
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct PendingEnvironmentMapLight {
+    diffuse: AssetId<Image>,
+    specular: AssetId<Image>,
+}
+
+#[derive(Resource, Default, Deref, DerefMut)]
+pub struct PendingEnvironmentMapLights(HashMap<PendingEnvironmentMapLight, Vec<Entity>>);
+
+pub fn extract_environment_maps(
+    mut environment_maps: ResMut<EnvironmentMapInstances>,
+    query: Extract<Query<(Entity, &EnvironmentMapLight)>>,
+    images: Res<RenderAssets<GpuImage>>,
+    mut pending_environment_map_images: ResMut<PendingEnvironmentMapLights>,
+) {
+    environment_maps.clear();
+    for (entity, environment_map_light) in &query {
+        let (Some(diffuse), Some(specular)) = (
+            images.get_key(environment_map_light.diffuse_map.id()),
+            images.get_key(environment_map_light.specular_map.id()),
+        ) else {
+            pending_environment_map_images
+                .entry(PendingEnvironmentMapLight {
+                    diffuse: environment_map_light.diffuse_map.id(),
+                    specular: environment_map_light.specular_map.id(),
+                })
+                .or_default()
+                .push(entity);
+            continue;
+        };
+        environment_maps.insert(entity, EnvironmentMapIds { diffuse, specular });
+    }
+}
 
-    fn extract(item: QueryItem<'_, Self::QueryData>) -> Option<Self> {
-        Some(EnvironmentMapIds {
-            diffuse: item.diffuse_map.id(),
-            specular: item.specular_map.id(),
-        })
+pub fn update_pending_environment_maps(
```

Comment:
I did want this to use the plugin, but I feel like it's an exception, and I don't want to make the trait for the helper plugin much more complicated just to cover this case. Not certain it's the right choice though.